### PR TITLE
Add device config page

### PR DESCRIPTION
**Pull request by bramkragten**
First step for changing the integration pages

image

### CODE REVIEW

**Comment by dmulcahey** [comment]
I may be over thinking this... but what do you think about leveraging a supplier concept that integrations can use to customize this? 

Ex: zigbee devices have additional attributes that are useful

- nwk address
- ieee address
- available status
- network link quality and signal strength

Also, not everything may live in the device registry. What if we added a way to supply the WS command to supply the devices (or one to provide additional data for each in the registry), list of attrs to show in the table and maybe even the row click action (think device more info) to individual integrations?

Disclaimer: I have a card that performs these functions for ZHA already. I'd love to trash it for this... My card is a mess and this is awesome! Plus it's core... no additional config for the users.

**Comment by frenck** [comment]
I really don't like these tables. There is less information and I feel like I'm looking at an Excel sheet.
I can handle change and applaud the change in general, however, I don't see a need for this in general.

For example, in the current view, I can see the entities, including the names of those entities, of a device. In this suggested view, I only get a list of icons. That will look hideous when a devices brings in a lot of entities with the same icon. There is no way to differentiate from that point.

I have a hard time to understand how this, from a UX perspective, is an improvement over what we already have.

**Comment by Kane610** [comment]
Awesome work Bram! But Im with frenck, it is clean and all, but you lose details and it is harder to approach. This requires another iteration

**Comment by bramkragten** [comment]
@dmulcahey Yes, that is something I was planning to do. We want to revamp the device and integration page and the zwave and zha panels should move to the integration settings.

As a next step we could look into getting specific integration settings in the device page and the integration page so everything is easy to find and at a logical place.

**Comment by bramkragten** [comment]
@frenck this is a new page, so it won’t replace anything and you won’t lose information.

The idea behind the data-table is that it scales. In the old UI if you have hundreds of devices if would be hard to find the one you want.

With a table you can sort and filter and easily find the device you search. If you click the device you get all the details you are used to and we will expand that in the future.

This is a MVP, and improvements will be required, we still have to determine what information has to be shown, like battery level, online status, etc.

**Comment by pvizeli** [comment]
I like that because it's looking clean, unlike the current view on integration page to see all entities.

**Comment by frenck** [comment]
@bramkragten It doesn't scale, that is my whole issue with it.

I know it is a new screen, however, it sets a presendece (apparently) for the future. Since this is a continuation of something happening already.

This is how one of the data tables look like on mobile:

IMG_5530

This is another example view with the "old"/"current" views:

IMG_5529

Please explain to me, how the data tables is an improvement in this case.

Secondly, I still would like to read up on the philosophy behind this all, could you please provide me a link @bramkragten? Thanks!

**Comment by bramkragten** [comment]
@frenck with scales, I mean that the layout also works with 1000 devices, and not only with 3 devices like the old layout.

As this screen will show all your devices, not just the devices from 1 integration like the screen you showed.

It is not a replacement of that screen, it is a overview of all your devices. If you would show that in the way the current integration layout does, it would be messy and hard to find the device you are looking for. 

The details of the device and all it entities is something that belongs on a device detail page IMHO, not in a overview of the devices.

There is no link to a philosophy, this is a ongoing process and this is the first idea that Paulus and I came up with. We evaluate how that works and what can be improved and work from there.

**Comment by frenck** [comment]
@bramkragten I know my screenshots are not reflecting this PR, since it is a new screen. I would not be able to provide screenshots for something that has not been build. Hence, I've provided screenshots of similar pages. So please, don't use that as an argument, since that makes no sense.

I know the current way of displaying has issues with displaying a large number of items as well, however, my point is, that the offered solution is not improving on it from a mobile perspective. Which it never will with data tables, as this is a general UX issue with those on mobile in general.

Hence, me asking for the philosophy behind it. If we, somewhere, decided on ditching UX for mobile, then I'm happy to back off.

**Comment by bramkragten** [comment]
Well like I explained on discord, these views (config) are mostly used on desktop. So that is the main focus at first. I agree that data tables won’t work on mobile and that we will need a completely different design for mobile. But that is something for the future as we can’t do everything at once.

**Comment by frenck** [comment]
The question that raises in that case: Are data tables the solution?

What other possibilities have we explored?

**Comment by bramkragten** [comment]
For mobile I would like it to eventually have a view like this: 
9C54EA03-F22C-4C6A-A8EE-A6CB1FBDF269

**Comment by frenck** [comment]
Agreed, Ubnt does that quite well, however, that screenshot shows a native app, in our case, we would indeed end up with tons of different views across our whole application, which is makes maintainability a bigger ish problem.

**Comment by arsaboo** [comment]
The proposed view is a huge improvement over our current one. It may not be perfect (especially on mobile), but I would take it any day. While we want to be optimized for mobile, I doubt that many of these backend tasks will be carried out on mobile.

**Comment by balloob** [comment]
It was partially my idea, so let me explain my thinking a bit. (disclaimer: It's 7AM here and I just had two full days of CEDIA, so my head is not 100% clear). Also we don't have a final mock-up of where we're headed, that would be a good idea. 

The idea is that we're going to introduce a new device management page in the config panel. It lists all devices and we can filter by area, config entry, name, etc. Clicking on a device will open a new device details page for full management of that device. This page would eventually also be opened when clicking on a device in the integrations page. 

Thinks that we want to make easier/possible:
 - Allow integrations without config flow to offer devices (and so also device automations). Right now integrations tab is the only entry into devices, there is no device registry page like we have area/entity/person).
 - Give users a full overview of the devices in their homes
 - See, and manage, device entities. (think renaming device + all entities at once)
 - Allow integrations to offer options for the device, like ZHA, Z-Wave. This will make this panel a 1-stop shop for device management.
 - Allow users to see in which device automations this device is involved
 - (future) allow users to upload photos of their devices and share those with the community

Here are some other UX thoughts:
 - When narrow, drop columns or make it look a list like ubnt. Maintenance wise it is not going to be much extra work, we have always had to test each view mobile/desktop, and I don't expect these views to change much.
 - Drop the entity icons, add a max width to the column to cut it off or replace with a number.

**Comment by SeanPM5** [comment]
Personally I think the current Integrations page and overall device experience could use a lot of work (I already wrote a detailed document with ideas & mockups that i shared with balloob a few months back). So I **love** that this particular aspect of Home Assistant is being given some attention, even if I'm not totally in love with this first iteration.

Some feedback and potential issues I see with this type of design (long comment incoming):

1. As frenck mentioned above, a table-based layout isn't mobile-friendly at all, which is the biggest drawback of this. Although mobile support is promised for later on, it's going to require different views or dropping columns which leads to an inconsistent user experience across device types. And until mobile support happens, it will be a poor UX in the meanwhile because:
    * It will be extremely jarring when a mobile user is comfortably navigating around Home Assistant and all of a sudden they unexpectedly get a page that is totally unusable on their device. 
    * The idea of 'you should use a desktop browser for this page, this page, and that page' is a slippery slope imo. Either a UI is entirely mobile-friendly or it's not, you can't really have something in-between like this without it feeling janky imo. 

2. Devices aren't grouped by integration, so in addition to causing way more horizontal scrolling it's also going to cause way more vertical scrolling too since every device now has its own entry / row. On the current Integrations page, my Google Cast entities only take up only one row since they're grouped together. On this design they would now take up a dozen lines. Now repeat that for every integration you have, and it could lead to hundreds of things being shown at once. I know this is kind of the whole point, but I feel like it would be information overload in a lot of cases.

3. Related to the above about devices no longer being grouped together, this means they're going to be located all over the place by default with the Device name ascending sorting. For example if you have Cast devices named "Bathroom Speaker", "Living Room Display", and "Zack's Chromecast," those would be located on the top, middle, and bottom of the list with dozens of completely unrelated things in-between. I realize there's an easy solution already baked into this design, to simply sort the table by Integration. But then that'd also be pretty much how it works today. 

4. Currently the three sort options that would be used 99% of the time (device name, integration, area) are on opposite sides of the table. The most-used controls should be next to each other imo.

5. The "Manufacturers" column is going to be problematic in many cases:
    * Not all devices have manufacturer info, so there would often be blank spaces like the Chromecast entries in that screenshot which doesn't look great.
    * Other times the manufacturer name can be ridiculously long which also doesn't look great and leads to more horizontal scrolling.
    * It could lead to user confusion - to give one example my Hue bridge has the manufacturer as "Signify" but all my other Hue devices are under "Philips" - these would not be listed together when sorted, even though they are very much related. 
    * Similar to above, what about companies that seemingly rebrand their products every other week. Then you'd have some devices under Alphabet Inc and others under Nest and others under Google. They wouldn't necessarily be together or where the user expects them to be.
    * But even ignoring the above points, I don't really think anybody typically browses their devices by manufacturer anyway? Is there an awesome use case I'm missing here?

6. The "Entities" column that displays the icons doesn't seem like it would scale well. In the example screenshot there's only 1-2 icons being displayed so it looks fine, but other integrations have many many more. The "Mobile App" (iOS 2.0 app) integration has 15-16 entities per device for example. So I can see three options of dealing with that, all of which aren't great:
    * Display them all, which would cause _even more_ horizontal scrolling
    * Only show a certain amount of icons before wrapping icons another line, but then you'd have some table rows that are 8x the height of others which would look hideous imo.
    * Or show only a certain amount of icons, in which case why show icons at all? Because it would be extremely frustrating if the entity you wanted didn't meet some arbitrary "first four icons" cutoff. IMO it should be all icons or no icons.

7. It's not exactly clear to me where this leaves the Integrations page in the end? It feels like there's going to be a lot of overlap between the two sections now. The users devices will show up on both pages, just with a different UI? If I want to add a new device to Home Assistant, do I do it from the "Devices" page or the "Integrations" page? This could get confusing very quickly if there's not a clear distinction.

I do really like the overall idea of a dedicated Devices page. I like the Area stuff, I like the search. And even though I kinda complained about it above, I could totally see how some people would like the "show me everything all at once" approach. There's some intriguing ideas here for sure.

But personally I think that focusing on the mobile experience _first_ would probably make more sense, because it forces you to think about what elements are truly essential, and that leads to a cleaner UI overall. For example, I don't believe that "Manufacturer" and "Model" is necessary information to display on this overview page given some of the problems I mentioned above, and especially after learning there's a "Device Details" page on click through.

**Comment by iantrich** [comment]
Can you share those ideas/mockups?

**Comment by SeanPM5** [comment]
Looking at it again it's much more focused on the overall Integrations experience rather than specifically Devices, and it's a couple months old now so a bunch of stuff is outdated, but [here](https://github.com/SeanPM5/homeassistant-config/wiki/Improving-Integrations).

**Comment by bramkragten** [comment]
I see that my ideas are not clear by just this screenshot, I will work on a final design in the next week that will clarify this and will be something we can work towards.

**Comment by bramkragten** [comment]
And I will definitely use parts of the designs of @SeanPM5 they are great!

**Comment by bramkragten** [comment]
Before I will start with my mockups let me explain a little what structure I have in mind so you can all comment on that:

# Integrations
Integrations will be our app store, you can search an integration, or see the integrations you have setup, it will not be focused on the devices but on the integration itself.

## Detail view
The detail page page of the integration will contain all information about that integration, like setting it up, settings and possibly events. This will make it easier for a user to know how an integration works and is setup, instead of having to search for it in the documentation.

### Settings
The detail page of the integration will also allow the user to change and view all the settings of the integration. This includes settings from the integration flow, like Deconz has, the system settings and specific settings like from Z-wave and ZHA.

### Devices
The third thing on the integration page would be the devices of that integration. It will give much less info than it does now, because most of the functions and info will be moved to the device page. Clicking on a device will get you to the device detail page like it would from the devices page.

# Devices
All devices will be listed like it is done in this PR, we should probably drop the manufacture, model and entities columns. The entities could be replaced by a badge with a number. Instead we want to add a picture of the device, the battery state if it has a battery sensor, and the state. Something like green/orange/red, green being all fine online, orange could be needs attention, update available or something, and red connection lost offline, doesn't work. The area would become a dropdown, so you can change the area for within the list (desktop only).

On mobile you will get the same information but not on a horizontal line but vertical, the sorting of the view will be move to a menu instead of header clicks.

## Detail view
If you click on a device you get to the device detail page. 

This page will show some basic information about the integration and a link to the integration to change settings of the integration.

It will show some specific information about the device if available, with the functions and possibilities in Home Assistant. It can also give advise on what you could do with the device in terms of automation. Especially since we now have device automations I think this could be helpful.

For instance, I have Zigbee push buttons now integrated with ZHA, all I see in my integration/device view now is a battery entity. A novice user would not have a clue how to use this device in an automation and that it should use events to capture clicks of the device.

If you click the suggestion you will be forwarded to the pre-filled automation editor.

### Settings
Of course the settings of a device will also be available, the area, name, and specific device settings.

The settings of ZHA, Z-wave and maybe other integrations that are now on a separate panel will also be present on this device page.

### Entities
All the entities of the device will also be listed with an easy interface to change the name and entity id and enable/disable the entity.

Also customisation of entities should be doable from the device page.

Eventually, I would love to have all integrations be device based, this would allow us to move things like entity registry and customisation completely. It would give us a clear structure:
Ingtegration -> Device -> Entity

So shoot away! Please share your opinion and suggestions.

**Comment by bramkragten** [comment]
For integrations that don't have devices, the entities section from the device detail page would be shown on the integration page.

**Comment by cogneato** [comment]
I've always appreciated @SeanPM5's breakdowns of previous UI elements that I've seen. I hadn't come across the integrations page until now. Good stuff! The design examples for integrations would also work well for the hassio addon store page, no? Two birds one stone?

> This will make it easier for a user to know how an integration works and is setup, instead of having to search for it in the documentation.

I see users search and end up with documentation that is incorrect or not appropriate for their install all the time, even when using the official pages. Pulling in more information away from "the docs" and straight into the UI when possible would go a long way.

**Comment by iantrich** [comment]
@cogneato relevant https://github.com/home-assistant/architecture/issues/282#issuecomment-529231097

**Comment by SeanPM5** [comment]
Those plans @bramkragten posted sound fantastic and really well thought out. Makes a very clear distinction between Integrations and Devices and there's lots of really user-friendly concepts in there. Nice!

I especially like the Device Details view suggesting how to use the device in automations etc. It's not too hard to get your phone added to Home Assistant these days, but then how do you use it for presence detection? How do you send notifications to it? The user is basically left to figure all this stuff out on their own after the initial setup process. So that idea of guiding them along the next steps until they have a working automation is incredibly helpful.

One suggestion I just thought of: this Devices UI makes it very easy to see which Areas devices are in and assign them, but we should probably consider ways of doing that with Persons also. Not all devices would have `device_tracker` entities, so a "Person" column probably wouldn't be suitable for the Overview page, but I think it would fit in on the Device Details page.

So on the Device Details page, maybe under the name of the device it could say something like "Belongs to Bram" (if it's assigned to `person.bram`) or "Assign Device to Person" if it's not. And clicking assign to person would bring up a person picker where you could easily associate that device with a Person.

File: src/panels/config/devices/ha-config-devices-dashboard.ts [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+            (device) => {
+              return {
+                device_name: device.name_by_user || device.name,
+                id: device.id,
+                manufacturer: device.manufacturer,
+                model: device.model,
+                area:
+                  !this.areas || !device || !device.area_id
+                    ? "No area"
+                    : this.areas.find(
```
Comment (balloob) [inline src/panels/config/devices/ha-config-devices-dashboard.ts]: I suggest that our memoization function create a data structure that gives us devices, but also already finds areas, integrations and entity ID representing the battery status (but not the actual fetching of state, as we should not memoize anything that requires `hass.states`, as that breaks caching).

File: src/components/ha-data-table.ts [hunk 2]
```diff
@@ -387,14 +395,21 @@ export class HaDataTable extends BaseElement {
...
 
     this._setRowChecked(rowId!, checkbox.checked);
     this.mdcFoundation.handleRowCheckboxChange(ev);
   }
 
+  private _handleRowClick(ev: Event) {
+    const rowId = (ev.target as HTMLElement)
+      .closest("tr")!
+      .getAttribute("data-row-id")!;
+    fireEvent(this, "row-click", {
```
Comment (balloob) [inline src/components/ha-data-table.ts]: I don't think that this one should bubble.

File: src/panels/config/devices/ha-config-device-page.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      : undefined;
+  }
+
+  protected firstUpdated(changedProps) {
+    super.firstUpdated(changedProps);
+    loadDeviceRegistryDetailDialog();
+  }
+
+  protected render() {
+    const device = this._device;
```
Comment (balloob) [inline src/panels/config/devices/ha-config-device-page.ts]: Should we memoize this ?

File: src/panels/config/devices/ha-config-device-page.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+          icon="hass:settings"
+          @click=${this._showSettings}
+        ></paper-icon-button>
+        <ha-device-card
+          .hass=${this.hass}
+          .areas=${this.areas}
+          .devices=${this.devices}
+          .device=${device}
+          .entities=${this.entities}
+          .hideSettings=${true}
```
Comment (balloob) [inline src/panels/config/devices/ha-config-device-page.ts]: Make sure it's type is set to boolean and you can just set a static attribute:

```suggestion
          hidesettings
```

File: src/panels/config/devices/ha-config-devices-dashboard.ts [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+
+import "../../../components/ha-card";
+import "../../../components/ha-data-table";
+import "../../../components/entity/ha-state-icon";
+import "../../../layouts/hass-subpage";
+import "../../../resources/ha-style";
+import "../../../components/ha-icon-next";
+
+import "../ha-config-section";
+// import { computeRTL } from "../../../common/util/compute_rtl";
```
Comment (balloob) [inline src/panels/config/devices/ha-config-devices-dashboard.ts]: comment

File: src/panels/config/devices/ha-config-devices-dashboard.ts [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+
+  private _devices = memoizeOne(
+    (
+      devices: DeviceRegistryEntry[],
+      entries: ConfigEntry[],
+      entities: EntityRegistryEntry[],
+      areas: AreaRegistryEntry[],
+      domain: string
+    ) => {
+      let outputDevices: DeviceRowData[] = [...devices];
```
Comment (balloob) [inline src/panels/config/devices/ha-config-devices-dashboard.ts]: Why do we copy it if we're not mutating  it?

File: src/panels/config/devices/ha-config-devices-dashboard.ts [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+        outputDevices = outputDevices.filter(
+          (device) =>
+            entries.find((entry) =>
+              device.config_entries.includes(entry.entry_id)
+            )!.domain === domain
+        );
+      }
+
+      outputDevices = outputDevices.map((device) => {
+        device.name = device.name_by_user || device.name;
```
Comment (balloob) [inline src/panels/config/devices/ha-config-devices-dashboard.ts]: You are mutatin data that you do not own. You should make a copy.

Comment (balloob) [reply src/panels/config/devices/ha-config-devices-dashboard.ts]: If you do `[...devices]`, you're only copying the array, the objects are not copied. (shallow copy)

Comment (balloob) [reply src/panels/config/devices/ha-config-devices-dashboard.ts]: Should we have a "<no name>" placeholder ?

Comment (balloob) [reply src/panels/config/devices/ha-config-devices-dashboard.ts]: I even wonder if we should extract that into a helper, we had the same problem in the device picker. Should also be localized.

File: src/panels/config/devices/ha-config-devices-dashboard.ts [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+            return data;
+          })}
+          @row-click=${this._handleRowClicked}
+        ></ha-data-table>
+      </hass-subpage>
+    `;
+  }
+
+  private _batteryEntity(device, entities): string | undefined {
+    const batteryEntity = entities.find(
```
Comment (balloob) [inline src/panels/config/devices/ha-config-devices-dashboard.ts]: I wonder if we should add helper `findEntityRegistryEntriesByDevice`

File: src/panels/config/devices/ha-config-devices.ts [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+import { UnsubscribeFunc } from "home-assistant-js-websocket";
+
+@customElement("ha-config-devices")
+class HaConfigDevices extends HassRouterPage {
+  @property() public hass!: HomeAssistant;
+  @property() public narrow!: boolean;
+
+  protected routerOptions: RouterOptions = {
+    defaultPage: "dashboard",
+    preloadAll: true,
```
Comment (balloob) [inline src/panels/config/devices/ha-config-devices.ts]: nothing to preload since there are no `load` functions  on the routes.

**Review by balloob — APPROVED** [review]
I think that this is a good first step. We'll make it mobile friendly in the next PR, this is already a beast as is. We'll keep iterating.

**Comment by frenck** [comment]
Thanks all for the discussion and bringing in clarity.
Good job @bramkragten, appreciated and sounds really sane 👍 
Oh and @SeanPM5 I really enjoyed the UX link you shared, really interesting!

File: src/panels/config/devices/ha-config-devices-dashboard.ts (hunk 7 above)
Comment (bramkragten) [reply src/panels/config/devices/ha-config-devices-dashboard.ts]: Helper is a good idea I think, let's add that in a separate PR

**Comment by elupus** [comment]
I ran into an issue here. config entries that only generate entities without unique id:s. will not be displayed as configured. arcam_fmj is one such component. It can't generate unique_id for it's entities under current rules.

It does still set device_info, but removing that doesn't help anyway.

**Pull request merged**